drivers/vhost/dsp-sof.c

```diff
@@ -0,0 +1,968 @@
+ *		Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>
+ */
+
+#include <asm/unaligned.h>
+
+#include <linux/device.h>
+#include <linux/file.h>
+#include <linux/firmware.h>
+#include <linux/fs.h>
+#include <linux/hw_random.h>
+#include <linux/miscdevice.h>
+#include <linux/module.h>
+#include <linux/slab.h>
+#include <linux/uio.h>
+#include <linux/vmalloc.h>
+#include <linux/workqueue.h>
+
+#include <sound/pcm_params.h>
+#include <sound/sof.h>
+
+#include <sound/sof/virtio.h>
+
+#include "dsp.h"
+
+#include "../sound/soc/sof/sof-priv.h"
+#include "../sound/soc/sof/ops.h"
+
+struct dsp_pipeline_connect {
+	struct list_head list;
+	int vq_idx;	/* VirtQ index */
+	int guest_id;	/* Guest component (virtual DAI) ID */
+	int host_id;	/* Host component (mixer) ID */
+	enum sof_ipc_stream_direction direction;
+};
+
+struct snd_sof_widget *dsp_sof_find_swidget_id(struct snd_sof_dev *sdev,
+					       unsigned int comp_id)
+{
+	struct snd_sof_widget *swidget;
+
+	list_for_each_entry(swidget, &sdev->widget_list, list)
+		if (swidget->comp_id == comp_id)
+			return swidget;
+
+	return NULL;
+}
+
+static struct snd_sof_dai *dsp_sof_find_dai_pipe(struct snd_sof_dev *sdev,
+						 unsigned int pipeline_id)
+{
+	struct snd_sof_dai *dai;
+
+	list_for_each_entry(dai, &sdev->dai_list, list)
+		if (dai->pipeline_id == pipeline_id)
+			return dai;
+
+	return NULL;
+}
+
+/*
+ * This function is used to find a BE substream. It uses the dai_link stream
+ * name for that. The current dai_link stream names are "vm_fe_playback" and
+ * "vm_fe_capture," which means only one Virtual Machine is supported and the VM
+ * only supports one playback pcm and one capture pcm. After we switch to the
+ * new topology, we can support multiple VMs and multiple PCM streams for each
+ * VM. This function may be abandoned after switching to the new topology.
+ */
+static struct snd_pcm_substream *dsp_sof_get_substream(struct snd_sof_dev *sdev,
+				struct snd_soc_pcm_runtime **rtd, int direction)
+{
+	struct snd_soc_card *card = sdev->card;
+	struct snd_soc_pcm_runtime *r;
+
+	for_each_card_rtds(card, r) {
+		struct snd_pcm_substream *substream;
+		struct snd_pcm_str *stream;
+		struct snd_pcm *pcm = r->pcm;
+		if (!pcm || !pcm->internal)
+			continue;
+
+		/*
+		 * We need to find a dedicated substream:
+		 * pcm->streams[dir].substream which is dedicated
+		 * used for vFE.
+		 */
+		stream = &pcm->streams[direction];
+		substream = stream->substream;
+		if (substream) {
+			struct snd_soc_dai_link *dai_link = r->dai_link;
+
+			/* FIXME: replace hard-coded stream name */
+			if (dai_link->stream_name &&
+			    (!strcmp(dai_link->stream_name, "vm_fe_playback") ||
+			     !strcmp(dai_link->stream_name, "vm_fe_capture"))) {
+				if (rtd)
+					*rtd = r;
+				return substream;
+			}
+		}
+	}
+
+	return NULL;
+}
+
+#define dsp_sof_find_spcm_comp sdev->core_ops.find_spcm_comp
+
+static int dsp_sof_assemble_params(struct sof_ipc_pcm_params *pcm,
```

Comment:
please comment all these function, what they do, why and when.

Reply:
comments added